sympy/sets/sets.py

```diff
@@ -994,6 +993,29 @@ def _eval_imageset(self, f):
             return
         var = f.variables[0]
 
+        if expr.is_Piecewise:
+            result = S.EmptySet
+            domain_set = self
+            for i in expr.args:
+                func = i[0]
+                if i[1] is S.true:
+                    intrvl = domain_set
+                else:
+                    intrvl = i[1].as_set()
+                    intrvl = Intersection(domain_set, intrvl)
+
+                if func.is_Number:
+                    image = FiniteSet(func)
+                else:
+                    image = imageset(Lambda(var, func), intrvl)
+                result = Union(result, image)
+
+                # remove the part which has been `imaged`
+                domain_set = Complement(domain_set, intrvl)
+                if domain_set.is_EmptySet:
+                    break
```

Comment:
I'm not sure how does line 1012-1015 helps, doesn't the fact that it is a `for` loop ensures that the loop will end? Are there cases when this statement will make the loop break before it supposed to break? Does it make any difference in the final result and are their any other advantages of doing things this way?